crates/ide_assists/src/handlers/promote_mod_file.rs

```diff
@@ -0,0 +1,162 @@
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    base_db::AnchoredPathBuf,
+};
+use syntax::{
+    ast::{self},
+    AstNode, TextRange,
+};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: promote_mod_file
+//
+// Moves inline module's contents to a separate file.
+//
+// ```
+// //- /main.rs
+// mod a;
+// //- /a.rs
+// $0fn t() {}
+// ```
+// ->
+// ```
+// fn t() {}
+// ```
+pub(crate) fn promote_mod_file(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let source_file = ctx.find_node_at_offset::<ast::SourceFile>()?;
+    let module = ctx.sema.to_module_def(ctx.frange.file_id)?;
+    if module.is_mod_rs(ctx.db()) {
+        return None;
+    }
+    let target = TextRange::new(
+        source_file.syntax().text_range().start(),
+        source_file.syntax().text_range().end(),
+    );
```

Comment:
I think making it triggerable for the whole file is way to noisy. Maybe make it triggerable on the first line or the `mod` item?